### PR TITLE
Flashlight simulation on map

### PR DESCRIPTION
**Pull request by voiperr**
## Purpose

Give players the ability to properly use NVG and flashlights on the map when it's dark. Add realistic flashlights to ACE.

## Demo:
https://www.youtube.com/watch?v=q8pVBif9KYU

##### ace_common:
* Adds: "visibleMapChanged" ACE eventhandler for when map opens/closes.

##### ace_map:
* Adds: Flashlight can be used on map, with behaviour analogous to real-life (flashlight beam scaling at zoom levels)
* Adds: Small amount of global illumination/beam effect when player is looking at map with flashlight

##### ace_flashlights (new):
* Adds: 3 flashlights: Fulton MX-991 (Blufor), KSF-1 (Opfor), Maglite XL50 (Indep/Civ)
* Adds: options for future flashlight classes of various colours: white, red, green, blue, yellow
* Changes: "acc_flashlight" made compatible with ace_map

##### ace_realisticnames:
* Adds: "acc_flashlight" has displayName "UTG Defender 126" (most visually similar real-world analogue found)

### CODE REVIEW

**Comment by alganthe** [comment]
:+1: very useful, and very intuitive.

File: addons/map/functions/fnc_getUnitFlashlights.sqf [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+#include "script_component.hpp"
+
+PARAMS_1(_unit);
+
+_flashlights = [];
+
+{
+    if (isText (configFile >> "CfgWeapons" >> _x >> "ItemInfo" >> "FlashLight" >> "ACE_Flashlight_Colour")) then {
+        if !(_x in _flashlights) then {
```
Comment (Glowbal) [inline addons/map/functions/fnc_getUnitFlashlights.sqf]: The above two if statements could/should be combined into one.

File: addons/map/stringtable.xml [hunk 2]
```diff
@@ -91,14 +91,10 @@
         <Key ID="STR_ACE_Map_BFT_Enabled_DisplayName">
             <English>BFT Enable</English>
             <Portuguese>RFA ativo</Portuguese>
-            <Polish>Aktywuj BFT</Polish>
-            <Czech>Povolit BFT</Czech>
```
Comment (Glowbal) [inline addons/map/stringtable.xml]: You removed some stringtable entries here

File: addons/map/stringtable.xml [hunk 3]
```diff
@@ -91,14 +91,10 @@
...
         </Key>
         <Key ID="STR_ACE_Map_BFT_Enabled_Description">
             <English>Enable Blue Force Tracking. Default: No</English>
             <Portuguese>Ativa Rastreio de Forças Azuis. Padrão: Não</Portuguese>
-            <Polish>Aktywuj Blue Force Tracking. Domyślnie: Nie</Polish>
-            <Czech>Povolit Blue Force Tracking. Výchozí: Ne</Czech>
```
Comment (Glowbal) [inline addons/map/stringtable.xml]: Here as well

File: addons/flashlights/CfgVehicles.hpp [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        class ACE_SelfActions {
+            //todo: add flashlight attach actions
+        }; 
+    };
+    
+    class NATO_Box_Base;
+    class EAST_Box_Base;
+    class IND_Box_Base;
+    class FIA_Box_Base_F;
+
```
Comment (Glowbal) [inline addons/flashlights/CfgVehicles.hpp]: You don't seem to have Item entries here. I think it would be a good idea to add those. For examples, look at the CfgVehicle file in the medical module and search for Item.

Comment (voiperr) [inline addons/flashlights/CfgVehicles.hpp]: Are we talking about the attach actions? Currently, they would serve no purpose, as I don't have the items set up as lightsources yet (will need help on that). I figured it would be better to leave them out until that's been made.

Comment (Glowbal) [inline addons/flashlights/CfgVehicles.hpp]: No. We are talking about the item entries. https://github.com/acemod/ACE3/blob/master/addons/medical/CfgVehicles.hpp#L746

Comment (voiperr) [inline addons/flashlights/CfgVehicles.hpp]: Gotcha. Forgot about Zeus/editor stuff. Added.

File: addons/map/stringtable.xml (hunk 3 above)
Comment (voiperr) [inline addons/map/stringtable.xml]: No idea how this happened; fixed.

Comment (Glowbal) [inline addons/map/stringtable.xml]: You added different text.

File: addons/flashlights/CfgVehicles.hpp (hunk 4 above)
Comment (bux578) [inline addons/flashlights/CfgVehicles.hpp]: Oh, and add it the ACE misc ammo box please (if you haven't already).

Comment (gienkov) [inline addons/flashlights/CfgVehicles.hpp]: May I ask for red light flashlight? Little feature request I know but very
useful. Is this doable?
9 lip 2015 11:58 "bux578" <notifications@github.com> napisał(a):

Comment (bux578) [inline addons/flashlights/CfgVehicles.hpp]: @gienkov did you watch the youtube vid?
https://www.youtube.com/watch?v=q8pVBif9KYU

**Comment by bux578** [comment]
Oh, and I guess we need some new models.

**Comment by gienkov** [comment]
That is so cool.

**Comment by bux578** [comment]
Really looking forward to it!

Oh and you've added the green light too :smile: 



I'd like to see an active flashlight emit some light around the unit using it, that'd be cool for PvP scenarios.

**Comment by gienkov** [comment]
It was like that in ACE2 bux so yeah its needed.

File: addons/flashlights/stringtable.xml [hunk 5]
```diff
@@ -0,0 +1,23 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project name="ACE">
+    <Package name="Flashlights">
+        <Key ID="STR_ACE_Flashlights_MX993_DisplayName">
+            <English>Fulton MX-993</English>
+        </Key>
```
Comment (gienkov) [inline addons/flashlights/stringtable.xml]: ```<Polish>Fulton MX-993</Polish>```

File: addons/flashlights/stringtable.xml [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+        <Key ID="STR_ACE_Flashlights_MX993_Description">
+            <English>Flashlight with red filter. For use on map.</English>
+        </Key>
```
Comment (gienkov) [inline addons/flashlights/stringtable.xml]: ```<Polish>Latarka z czerwonym filtrem. Używana do podświetlania mapy.</Polish>```

File: addons/flashlights/stringtable.xml [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+        <Key ID="STR_ACE_Flashlights_XL50_DisplayName">
+            <English>Maglite XL50</English>
+        </Key>
```
Comment (gienkov) [inline addons/flashlights/stringtable.xml]: ```<Polish>Maglite XL50</Polish>```

File: addons/flashlights/stringtable.xml [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+        <Key ID="STR_ACE_Flashlights_XL50_Description">
+            <English>White mini flashlight. For use on map.</English>
+        </Key>
```
Comment (gienkov) [inline addons/flashlights/stringtable.xml]: ```<Polish>Mini latarka. Światło białe. Używana do podświetlania mapy.</Polish>```

File: addons/flashlights/stringtable.xml [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+        <Key ID="STR_ACE_Flashlights_KSF1_DisplayName">
+            <English>KSF-1</English>
+        </Key>
```
Comment (gienkov) [inline addons/flashlights/stringtable.xml]: ```<Polish>KSF-1</Polish>```

File: addons/flashlights/stringtable.xml [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+        <Key ID="STR_ACE_Flashlights_KSF1_Description">
+            <English>Flashlight with red filter. For use on map.</English>
+        </Key>
```
Comment (gienkov) [inline addons/flashlights/stringtable.xml]: ```<Polish>Latarka z czerwonym filtrem. Używana do podświetlania mapy.</Polish>```

File: addons/map/stringtable.xml [hunk 11]
```diff
@@ -139,5 +139,23 @@
             <Czech>Umožňuje sledovat přátelské jednokty na mapě v rámci BFT.</Czech>
             <Portuguese>Permite que você acompanhe as posições no mapa das unidades aliadas com marcadores RFA.</Portuguese>
         </Key>
+        <Key ID="STR_ACE_Map_Action_Flashlights">
+            <English>Flashlights</English>
+        </Key>
```
Comment (gienkov) [inline addons/map/stringtable.xml]: ```<Polish>Latarki</Polish>```

File: addons/map/stringtable.xml [hunk 12]
```diff
@@ -139,5 +139,23 @@
...
+        <Key ID="STR_ACE_Map_Action_NVG">
+            <English>NVG</English>
+        </Key>
```
Comment (gienkov) [inline addons/map/stringtable.xml]: ```<Polish>Noktowizja </Polish>```
Without space

File: addons/map/stringtable.xml [hunk 13]
```diff
@@ -139,5 +139,23 @@
...
+        <Key ID="STR_ACE_Map_Action_NVGOn">
+            <English>On</English>
+        </Key>
```
Comment (gienkov) [inline addons/map/stringtable.xml]: ```<Polish>Włącz </Polish>```
Without space

File: addons/map/stringtable.xml [hunk 14]
```diff
@@ -139,5 +139,23 @@
...
+        <Key ID="STR_ACE_Map_Action_NVGOff">
+            <English>Off</English>
+        </Key>
```
Comment (gienkov) [inline addons/map/stringtable.xml]: ```<Polish>Wyłącz </Polish>```
Without space

File: addons/map/stringtable.xml [hunk 15]
```diff
@@ -139,5 +139,23 @@
...
+        <Key ID="STR_ACE_Map_Action_NVGUp">
+            <English>Increase Brightness</English>
+        </Key>
```
Comment (gienkov) [inline addons/map/stringtable.xml]: ```<Polish>Zwiększ czułość</Polish>```

File: addons/map/stringtable.xml [hunk 16]
```diff
@@ -139,5 +139,23 @@
...
+        <Key ID="STR_ACE_Map_Action_NVGDown">
+            <English>Decrease Brightness</English>
+        </Key>
```
Comment (gienkov) [inline addons/map/stringtable.xml]: ```<Polish>Zmniejsz czułość</Polish>```

**Comment by gienkov** [comment]
I will be grateful if you add translations I writed above.
Thanks

**Comment by voiperr** [comment]
External illumination:

External Illum

**Comment by ViperMaul** [comment]
Status?

**Comment by voiperr** [comment]
Functionality/coding's done, waiting on models (third person, for ground containers only). Imo they're not essential atm.

**Comment by voiperr** [comment]
Fairly large bug found; fixing.

**Comment by voiperr** [comment]
Should be good to go unless more bugs are found.

**Comment by alganthe** [comment]
:+1:

**Comment by Panimala** [comment]
Totally support this. Fan of the original mod.

**Comment by SilentSpike** [comment]
@acemod/owners I think this should be merged sooner rather than later so that there's plenty of testing time (assuming it's a wanted feature - I certainly think it looks great)

**Comment by Glowbal** [comment]
@voiperr Is it possible to add an enable toggle setting?

**Comment by voiperr** [comment]
Of course. Should it be dependent on mapIllumination, or do we want 3 separate modes? Should NVG illum and flashlight be separate?

**Comment by Glowbal** [comment]
I think one setting for the whole thing and another for toggling the flashlight glow would be sufficient? 

Default enabled for both, I'd say.

**Comment by voiperr** [comment]
More illumination settings added.

**Comment by bux578** [comment]
But flashlights are always active in "basic" and "advanced"? Imho, "advanced" should be the default.

**Comment by voiperr** [comment]
The flashlights do nothing in basic mode. I'm hesistant to make advanced default unless we arm most unit classes with flashlights by default. Otherwise, mass confusion on existing night missions.

**Comment by bux578** [comment]
ATM the map turns black if you're in a night mission and you can't see anything. So I don't think there will any more confusion. (And something that bugs me since we first introduced map illumination: IRL you're actually not really able to read a map with NVGs :smile:)

**Comment by voiperr** [comment]
For some reason I thought the map never turned fully dark currently. Welp, I'm not entirely sure what to make a setting then, because there seems to be a want for toggling the external glow(?).

Re NVG on map: I've heard conflicting reports on this. If someone could provide definitive knowledge...

**Comment by bux578** [comment]
> I've heard conflicting reports on this. If someone could provide definitive knowledge...

An acquaintance from the Bundeswehr (Airborne brigade) told me this. I don't have first hand knowledge, though.

**Comment by voiperr** [comment]
After various discussions, it was decided that the current implementation of NVG functionality on map was not realistic. Until proper blur functionality can be attained on the map display, this feature cannot be achieved realistically.

MapIllumination is reverted to either on or off. External glow is now a separate module option.

Unless there are other considerations, I think this is finally ready.

**Comment by Glowbal** [comment]
I notice some commits on here that should not be in this PR.

**Comment by voiperr** [comment]
Should be fixed (grrr github).

File: addons/map/functions/fnc_flashlightGlow.sqf [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+
+params ["_flashlight"];
+
+_light = GVAR(glow);
+if (!isNull _light) then {deleteVehicle _light};
+
+if (_flashlight != "") then {
+    _colour = getText (configFile >> "CfgWeapons" >> _flashlight >> "ItemInfo" >> "FlashLight" >> "ACE_Flashlight_Colour");
+
+    _class = switch (_colour) do {
```
Comment (Glowbal) [inline addons/map/functions/fnc_flashlightGlow.sqf]: 1. I'd suggest adding either a default value or some error handling in case someone uses the wrong color input.
2. Switch is case sensitive. Maybe add an `toLower` here?

File: addons/map/functions/fnc_getUnitFlashlights.sqf [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+
+#include "script_component.hpp"
+
+params ["_unit"];
+
+_flashlights = [];
+
+{
+    if ((isText (configFile >> "CfgWeapons" >> _x >> "ItemInfo" >> "FlashLight" >> "ACE_Flashlight_Colour")) && !(_x in _flashlights)) then {
+            _flashlights pushBack _x;
```
Comment (Glowbal) [inline addons/map/functions/fnc_getUnitFlashlights.sqf]: Indentation is wrong

File: addons/map/CfgVehicles.hpp [hunk 19]
```diff
@@ -39,13 +61,14 @@ class CfgVehicles {
         };
     };
 
-    class ACE_ModuleBlueForceTracking: ACE_Module {
+    class Module_F;
+    class ACE_ModuleBlueForceTracking: Module_F {
```
Comment (Glowbal) [inline addons/map/CfgVehicles.hpp]: Why is this changed in the PR for flashlights?

File: addons/common/XEH_postInit.sqf [hunk 20]
```diff
@@ -157,7 +158,7 @@ call FUNC(checkFiles);
     //Event that settings are safe to use:
     ["SettingsInitialized", []] call FUNC(localEvent);
 
-}, 0, [false]] call CBA_fnc_addPerFrameHandler;
+}, 0, [false]] call cba_fnc_addPerFrameHandler;
```
Comment (Glowbal) [inline addons/common/XEH_postInit.sqf]: The standard is having `CBA` upper case afaik. @ViperMaul  ?

Comment (jonpas) [inline addons/common/XEH_postInit.sqf]: Correct.

**Comment by voiperr** [comment]
Implemented suggested changes.

File: addons/map/functions/fnc_flashlightGlow.sqf [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+
+params ["_flashlight"];
+
+_light = GVAR(glow);
+if (!isNull _light) then {deleteVehicle _light};
+
+if (_flashlight != "") then {
+    _colour = getText (configFile >> "CfgWeapons" >> _flashlight >> "ItemInfo" >> "FlashLight" >> "ACE_Flashlight_Colour");
+
+    _class = switch (toLower _colour) do {
```
Comment (jokoho48) [inline addons/map/functions/fnc_flashlightGlow.sqf]: why not useing a format?
_class = format ["ACE_FlashlightProxy_%1",_color];

**Comment by Glowbal** [comment]
:+1: Thanks @voiperr
